core/pkg/server/sender.go

```diff
@@ -157,13 +165,47 @@ func NewSender(
 		}
 		maps.Copy(graphqlHeaders, settings.GetXExtraHttpHeaders().GetValue())
 
+		// Hook to capture response body for internal printer (for now)
+		// This is going to be removed once we refactor our loggers.
+		// We should just use the logger to capture this information.
+		repsonseHook := func(_ retryablehttp.Logger, resp *http.Response) {
+
+			if resp != nil && resp.Body != nil && resp.StatusCode != http.StatusOK {
+				// Read the response body
+				bodyContent, err := io.ReadAll(resp.Body)
+				if err != nil {
+					logger.Error("sender: failed to read response body", "error", err)
+					return
+				}
+				// Close the original body
+				resp.Body.Close()
+
+				// Use the body content in the internal printer
+				//
+				// Note that message is an empty string, since we need to
+				// comply with the interface of the Logger
+				// This will allow us to capture the response body.
+				// Note: this is going to be removed once we refactor our
+				// loggers. We should capture this information in the logger
+				sender.internalPrinter.Printf("", &service.HttpResponse{
+					HttpResponseText: string(bodyContent),
+					HttpStatusCode:   int32(resp.StatusCode),
+				},
+				)
+
+				// Restore the response body so it can be used again
+				resp.Body = io.NopCloser(bytes.NewReader(bodyContent))
```

Comment:
Should this be done in a `defer` immediately after `io.ReadAll(resp.Body)`?

Reply:
yeah you are right...

Reply:
maybe it should even be immediate rather than deferred; i didn't think about it too hard